File: src/day06.py


with open("input/day06", "r") as data:
    current_set = set()
    sum_of_set_sizes = 0
    for line in data:
        line = line.strip()
        if len(line) == 0:
            sum_of_set_sizes += len(current_set)
            current_set = set()
        else:
            current_set = current_set.union(set(line))

    print("Part 1: ", sum_of_set_sizes)


with open("input/day06", "r") as data:
    current_set = None
    sum_of_set_sizes = 0
    for line in data:
        line = line.strip()
        if len(line) == 0:
            sum_of_set_sizes += len(current_set)
            current_set = None
        else:
            if current_set is None:
                current_set = set(line)
            else:
                current_set = current_set.intersection(set(line))  # pt2

    print("Part 2: ", sum_of_set_sizes)

File: src/day06.ts
import { read_empty_line_and_whitespace_separated_strings } from "./read_file_utils";
import { intersect } from "./set_utils";



export function characters_in_any_of(strings: string[]): Set<string> {
    const sets: Array<Set<string>> = strings.map((value: string) => new Set(value));
    const union = sets.reduce((accumulation: Set<string>, currentValue: Set<string>) => accumulation = new Set([...accumulation, ...currentValue]));
    return union;
}

export function characters_in_all_of(strings: string[]): Set<string> {
    const sets: Array<Set<string>> = strings.map((value: string) => new Set(value));
    const union = sets.reduce((accumulation: Set<string>, currentValue: Set<string>) => accumulation = intersect(accumulation, currentValue));
    return union;
}


if (require.main === module) {
    const input = read_empty_line_and_whitespace_separated_strings("input/day06");

    {
        let sum_of_set_sizes = 0;
        input.forEach(element => sum_of_set_sizes += characters_in_any_of(element).size);
        console.log("Part 1: ", sum_of_set_sizes);
    }

    {
        let sum_of_set_sizes = 0;
        input.forEach(element => sum_of_set_sizes += characters_in_all_of(element).size);
        console.log("Part 2: ", sum_of_set_sizes);
    }
}